src/strands/multiagent/graph.py

```diff
@@ -0,0 +1,483 @@
+- Output propagation along edges
+- Topological sort for execution ordering
+- Clear dependency management
+- Supports nested graphs (Graph as a node in another Graph)
+"""
+
+import asyncio
+import logging
+import time
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass, field
+from typing import Any, Callable, Tuple, cast
+
+from ..agent import Agent, AgentResult
+from ..types.event_loop import Metrics, Usage
+from .base import MultiAgentBase, MultiAgentResult, NodeResult, Status
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class GraphState:
+    """Graph execution state.
+
+    Attributes:
+        status: Current execution status of the graph.
+        completed_nodes: Set of nodes that have completed execution.
+        failed_nodes: Set of nodes that failed during execution.
+        execution_order: List of nodes in the order they were executed.
+        task: The original input prompt/query provided to the graph execution.
+              This represents the actual work to be performed by the graph as a whole.
+              Entry point nodes receive this task as their input if they have no dependencies.
+    """
+
+    # Execution state
+    status: Status = Status.PENDING
+    completed_nodes: set["GraphNode"] = field(default_factory=set)
+    failed_nodes: set["GraphNode"] = field(default_factory=set)
+    execution_order: list["GraphNode"] = field(default_factory=list)
+    task: str = ""
+
+    # Results
+    results: dict[str, NodeResult] = field(default_factory=dict)
+
+    # Accumulated metrics
+    accumulated_usage: Usage = field(default_factory=lambda: Usage(inputTokens=0, outputTokens=0, totalTokens=0))
+    accumulated_metrics: Metrics = field(default_factory=lambda: Metrics(latencyMs=0))
+    execution_count: int = 0
+    execution_time: int = 0
+
+    # Graph structure info
+    total_nodes: int = 0
+    edges: list[Tuple["GraphNode", "GraphNode"]] = field(default_factory=list)
+    entry_points: list["GraphNode"] = field(default_factory=list)
+
+
+@dataclass
+class GraphResult(MultiAgentResult):
+    """Result from graph execution - extends MultiAgentResult with graph-specific details.
+
+    The status field represents the outcome of the graph execution:
+    - COMPLETED: The graph execution was successfully accomplished
+    - FAILED: The graph execution failed or produced an error
+    """
+
+    status: Status = Status.PENDING
+    total_nodes: int = 0
+    completed_nodes: int = 0
+    failed_nodes: int = 0
+    execution_order: list["GraphNode"] = field(default_factory=list)
+    edges: list[Tuple["GraphNode", "GraphNode"]] = field(default_factory=list)
+    entry_points: list["GraphNode"] = field(default_factory=list)
+
+
+@dataclass
+class GraphEdge:
+    """Represents an edge in the graph with an optional condition."""
+
+    from_node: "GraphNode"
+    to_node: "GraphNode"
+    condition: Callable[[GraphState], bool] | None = None
```

Comment:
nit for consistency with the SDK, `Optional[Callable[[GraphState], bool]]` (I'd like the whole SDK converted to use this syntax but until then)